src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs

```diff
@@ -0,0 +1,1029 @@
+                algorithmIdentifier.Parameters.Value,
+                AsnEncodingRules.BER);
+
+            int iterationCount = ClampIterationCount(pbeParameters.IterationCount);
+            Span<byte> iv = stackalloc byte[cipher.BlockSize / 8];
+            Span<byte> key = stackalloc byte[cipher.KeySize / 8];
+            ReadOnlySpan<byte> saltSpan = pbeParameters.Salt.Span;
+
+            try
+            {
+                Pkcs12Kdf.DeriveIV(
+                    password,
+                    hashAlgorithm,
+                    iterationCount,
+                    saltSpan,
+                    iv);
+
+                Pkcs12Kdf.DeriveCipherKey(
+                    password,
+                    hashAlgorithm,
+                    iterationCount,
+                    saltSpan,
+                    key);
+
+                return Decrypt(cipher, key, iv, encryptedData, destination);
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(key);
+                CryptographicOperations.ZeroMemory(iv);
+            }
+        }
+
+        private static unsafe int Decrypt(
+            SymmetricAlgorithm cipher,
+            ReadOnlySpan<byte> key,
+            ReadOnlySpan<byte> iv,
+            ReadOnlySpan<byte> encryptedData,
+            Span<byte> destination)
+        {
+            // When we define a Span-based decryption API this should be changed to use it.
+            byte[] tmpKey = new byte[key.Length];
+            byte[] tmpIv = new byte[iv.Length];
+            byte[] rentedEncryptedData = ArrayPool.Rent(encryptedData.Length);
+            byte[] rentedDestination = ArrayPool.Rent(destination.Length);
+
+            // Keep all the arrays pinned so they can be correctly cleared
+            fixed (byte* tmpKeyPtr = tmpKey)
+            fixed (byte* tmpIvPtr = tmpIv)
+            fixed (byte* rentedEncryptedDataPtr = rentedEncryptedData)
+            fixed (byte* rentedDestinationPtr = rentedDestination)
+            {
+                try
+                {
+                    key.CopyTo(tmpKey);
+                    iv.CopyTo(tmpIv);
+
+                    using (ICryptoTransform decryptor = cipher.CreateDecryptor(tmpKey, tmpIv))
+                    {
+                        Debug.Assert(decryptor.CanTransformMultipleBlocks);
+
+                        encryptedData.CopyTo(rentedEncryptedData);
+
+                        int writeOffset = decryptor.TransformBlock(
+                            rentedEncryptedData,
+                            0,
+                            encryptedData.Length,
+                            rentedDestination,
+                            0);
+
+                        rentedDestination.AsSpan(0, writeOffset).CopyTo(destination);
+
+                        byte[] tmpEnd = decryptor.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+
+                        fixed (byte* tmpEndPtr = tmpEnd)
```

Comment:
Is this just defense in depth?  There's a window here where tmpEnd could be moved around the heap, leaving shadows that won't be cleared.

Reply:
yes, we do it for as long as we can but we currently don't have ideal solution